Feature/token service

erikwiffin opened the pull request:
Fixes https://github.com/sync-for-science/tracking/issues/8.

erikwiffin committed:
Initial 'none' implementation of TokenService

erikwiffin committed:
Store access tokens in local sqlite db

erikwiffin committed:
Refactoring token generation

- Separate authorization code and authorization token functionality.
- Use eslint for consistent styling.
- Use more promises in token service, create a promise wrapper for sqlite calls.

erikwiffin committed:
Properly handle refresh tokens

erikwiffin committed:
Properly handling revoking refresh tokens

jmandel commented on src/services/database.js:
```diff
@@ -0,0 +1,56 @@
+
+      return new Promise(function (resolve, reject) {
+        args.push(function (err) {
+          (err) ? reject(err) : resolve(this);
```
What does `this` refer to right here?

erikwiffin replied:
From the sqlite3 docs (https://github.com/mapbox/node-sqlite3/wiki/API#databaserunsql-param--callback) it was intended to be the statement object of the query, but I'm pretty sure I haven't tested it, and looking at it now `this` probably got rebound once inside the Promise. I think I'll change that to `resolve()` since nothing is using that statement object.

jmandel commented on src/services/token.js:
```diff
@@ -0,0 +1,196 @@
+
+        switch (row["token_type"]) {
+        case "access_token":
+          query = "DELETE FROM access_token WHERE access_token = ?";
```
I wonder if we should update the rows by setting revokedAt=now(), or something similar, rather than deleting. How much chaos would that cause?

jmandel replied:
I should add that making the rows disappear is fine too -- but we want some kind of persistent log that tells the story of what happened.

erikwiffin replied:
I didn't realize you wanted a persistent log of expired/revoked access_tokens. That actually negates my intended use of Redis and the ability to have keys expire and disappear on their own. Setting revokedAt=now() should be easy enough.

jmandel replied:
Well, Redis would be fine for storing _active_ tokens, as long as we had the discipline of a log. In a real system that would be reasonable design; here, having two persistence stores would probably be over-engineering :-)

jmandel replied:
And my apologies for not articulating this along with requirements: it only occurred to me explicitly when I saw the approach you'd taken to deleting rows.

jmandel commented on src/smart/auth.js:
```diff
@@ -21,13 +19,13 @@ var lookups = [
   * Create an authorization code. This is for a trusted static front-end to call,
   * as indicated by the "smart/portal" scope. In general, only "we" call this;
-  * other apps don't.
+  * other apps don"t.
```
What's up with this?  ;)

erikwiffin replied:
`s/'/"/g` to get all the quotes to match. :)

jmandel commented:
This is lovely - thanks for your attention to detail, and especially for applying sane linting and standardizing styles. 

Will you take a look at the failing test in https://travis-ci.org/smart-on-fhir/smart-stub/builds/126518197 ? My guess is that the test expects to be able to simply sign a JWT that looks legit to the server, but with the new  DB backend this is no longer enough (because the JWT won't have appeared in the db). We might just get rid of the offending test if it doesn't make sense anymore, or we could ensure the offending test runs in "naive" token mode.

erikwiffin committed:
Respond to code review feedback.

- Use `revoked_at` instead of deleting revoked tokens.
- Removed failing and no longer relevant integration test.

jmandel commented:
Awesome. Squashed and merged. (If that was premature, no worries -- just let me know.)
